docs/guides/ssr.md

```diff
@@ -22,9 +20,214 @@ The most basic server rendering in React Router is pretty straightforward. Howev
+  let url = new URL(req.originalUrl || req.url, origin);
+
+  let controller = new AbortController();
+  req.on("close", () => controller.abort());
+
+  let headers = new Headers();
+
+  for (let [key, values] of Object.entries(req.headers)) {
+    if (values) {
+      if (Array.isArray(values)) {
+        for (let value of values) {
+          headers.append(key, value);
+        }
+      } else {
+        headers.set(key, values);
+      }
+    }
+  }
+
+  let init = {
+    method: req.method,
+    headers,
+    signal: controller.signal,
+  };
+
+  if (req.method !== "GET" && req.method !== "HEAD") {
+    init.body = req.body;
+  }
+
+  return new Request(url.href, init);
+};
+```
+
+Once we've loaded our data by executing all of the matched route loaders for the incoming request, we use `createStaticRouter()` and `<StaticRouterProvider>` to render the HTML and send a response back to the browser:
+
+```js filename=server.jsx lines=[5-16]
+app.get("*", async (req, res) => {
+  let fetchRequest = createFetchRequest(req);
+  let context = await handler.query(fetchRequest);
+
+  let router = createStaticRouter(
+    handler.dataRoutes,
+    context
+  );
+  let html = ReactDOMServer.renderToString(
+    <StaticRouterProvider
+      router={router}
+      context={context}
+    />
+  );
+
+  res.send("<!DOCTYPE html>" + html);
+});
+```
+
+Once we've sent the HTML back to the browser, we'll need to "hydrate" the application on the client using `createBrowserRouter()` and `<RouterProvider>`:
+
+```jsx filename=entry-client.jsx lines=[10-15]
+import * as React from "react";
+import ReactDOM from "react-dom/client";
+import {
+  createBrowserRouter,
+  RouterProvider,
+} from "react-router-dom";
+
+import { routes } from "./routes";
+
+let router = createBrowserRouter(routes);
+
+ReactDOM.hydrateRoot(
+  document.getElementById("app"),
+  <RouterProvider router={router} />
+);
+```
+
+And with that you've got a server-side-rendered and hydrated application! For a working example, you may also refer to the [example][ssr-data-router-example] in the Github repository.
+
+### Additional Concepts
+
+As mentioned above, server-side rendering is tricky at scale and for production-grade applications, and we strongly recommend checking out [Remix][remix] if that's your goal. But if you are going the manual route, here's a few additional concepts you may need to consider:
+
+#### Redirects
+
+If any loaders redirect, `handler.query` will return the `Response` directly so you should check that and send a redirect response instead of attempting to render an HTML document:
+
+```js filename=server.jsx lines=[5-10]
+app.get("*", async (req, res) => {
+  let fetchRequest = createFetchRequest(req);
+  let context = await handler.query(fetchRequest);
+
+  if (
+    context instanceof Response &&
+    (context.status === 301 || context.status === 302)
+  ) {
+    return res.redirect(
+      e.status,
+      e.headers.get("Location")
+    );
```

Comment:
i think we want `context` here?

```suggestion
    return res.redirect(
      context.status,
      context.headers.get("Location")
    );
```

Reply:
haha yep!  I literally just saw and fixed these 2 same typos!